orangecontrib/text/widgets/utils/concurrent.py

```diff
@@ -60,69 +144,60 @@ def join(self):
         if self._thread is not None and self._thread.is_alive():
             self._thread.join()
 
-    def progressBarSet(self, value, processEvents=None):
-        """ Changes default processEvents value. """
-        super().progressBarSet(value, processEvents=processEvents)
+    def should_break(self):
+        return not self.running
 
 
-def optional_args(fn):
-    """
-    Decorator used for decorating other decorators to enable optional arguments.
-    I.e. the other decorator can then be called either as `@wrap` or as `@wrap(args)`.
+class AsyncMethod(QObject):
+    def __init__(self, method):
+        super().__init__()
+        self.method = method
+        self.method_name = method.__name__
+        self.finish_callback = None
+        self.start_callback = None
 
-    Adopted from: http://stackoverflow.com/a/20966822/892987
-    """
-    def wrapped_decorator(*args, **kwargs):
-        if len(args) == 1 and callable(args[0]):
-            return fn(args[0])
-        else:
-            def real_decorator(decorate):
-                return fn(decorate, *args, **kwargs)
-            return real_decorator
+    def __get__(self, instance, owner):
+        """ Bounds methods with instance. """
+        bounded = BoundAsyncMethod(self, instance)
+        setattr(instance, self.method.__name__, bounded)
+        return bounded
 
-    return wrapped_decorator
+    def on_start(self, callback):
+        self.start_callback = callback.__name__
+        return Slot()(callback)
 
+    def callback(self, method):
```

Comment:
I'm not sure here. Even thought `callback` is only used  for progress reporting it can be used elsewhere.  
For instance, it's possible to use `callback` instead of `on_start` and manually run it in the beginning of the main method.